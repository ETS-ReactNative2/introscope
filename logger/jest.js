if (
    global.expect &&
    global.expect.addSnapshotSerializer &&
    !process.env.__EFFECTS_LOGGER__NO_JEST_INTERGATION
) {
    const { spySnapshotSerializer } = require('./spySnapshotSerializer');
    const {
        effectsLogSnapshotSerializer,
    } = require('./effectsLogSnapshotSerializer');

    global.expect.addSnapshotSerializer(spySnapshotSerializer);
    global.expect.addSnapshotSerializer(effectsLogSnapshotSerializer);
}
