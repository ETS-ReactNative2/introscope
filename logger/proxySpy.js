const { cloneDeepWith } = require('lodash');

const spiesRegistry = new WeakMap();

const putToRegistry = (spy, target) => {
    spiesRegistry.set(spy, target);
    return spy;
};

const isSpy = spy => spiesRegistry.has(spy);
const getSpyTarget = spy => spiesRegistry.get(spy);

const serializeWithSpies = v =>
    cloneDeepWith(v, value => (isSpy(value) ? getSpyTarget(value) : value));

const isSpialbe = v =>
    typeof v == 'function' || (typeof v == 'object' && v !== null);

const proxySpyFactory = ({ serialize }) => (log, id, v) =>
    isSpialbe(v)
        ? putToRegistry(
              new Proxy(v, {
                  apply(_, that, args) {
                      if (that === undefined) {
                          log(['call', id, serialize(args)]);
                      } else {
                          log(['apply', id, serialize(that), serialize(args)]);
                      }
                      return Reflect.apply(...arguments);
                  },
                  get(_, prop) {
                      log(['get', id, prop]);
                      return Reflect.get(...arguments);
                  },
                  set(_, prop, value) {
                      log(['set', id, prop, serialize(value)]);
                      return Reflect.set(...arguments);
                  },
              }),
              v,
          )
        : v;

const proxySpy = proxySpyFactory({ serialize: serializeWithSpies });

module.exports = {
    serializeWithSpies,
    isSpy,
    getSpyTarget,
    proxySpyFactory,
    proxySpy,
};
