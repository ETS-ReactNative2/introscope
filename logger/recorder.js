const { readFileSync, writeFileSync } = require('fs');

expect.extend({
    __getExpectContext(_, cb) {
        cb(this);
        return {
            pass: true,
            message: '',
        };
    },
});

class Recorder {
    constructor(resultsFile, mode) {
        this.resultsFile = resultsFile;

        if (mode == 'playback') {
            try {
                const results = JSON.parse(readFileSync(this.resultsFile));
                this.playback = true;
                this.results = results;
            } catch (_) {}
            return;
        }

        if (mode == 'record') {
            this.recording = true;
            this.results = [];
            return;
        }
    }

    save() {
        if (!this.recording) return;

        writeFileSync(this.resultsFile, JSON.stringify(this.results, null, 2));
    }
}

const getRecorder = () => {
    let context;
    expect(null).__getExpectContext(c => (context = c));

    const filename = context.snapshotState._snapshotPath.replace(
        /snap$/,
        'record',
    );
    const mode =
        context.snapshotState._updateSnapshot === 'all' ? 'record' : 'playback';

    return new Recorder(filename, mode);
};

module.exports = {
    getRecorder,
};
