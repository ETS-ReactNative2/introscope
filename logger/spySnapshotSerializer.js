const { isSpy, getSpyTarget } = require('./proxySpy');

const spySnapshotSerializer = {
    test(val) {
        return isSpy(val);
    },
    serialize(val, config, indentation, depth, refs, printer) {
        return printer(
            getSpyTarget(val),
            config,
            indentation,
            depth,
            refs,
            printer,
        );
    },
};
module.exports = {
    spySnapshotSerializer,
};
