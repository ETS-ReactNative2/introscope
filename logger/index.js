const { proxySpy } = require('./proxySpy');

const effectsLoggerLogRegistry = new WeakMap();
const isEffectsLoggerLog = log => effectsLoggerLogRegistry.has(log);

const newLog = () => {
    const log = [];
    effectsLoggerLogRegistry.set(log, true);
    return log;
};

// from https://github.com/peter-leonov/test-plan
const functionMocker = (log = newLog()) =>
    new Proxy(() => log, {
        get(_, prop) {
            return v =>
                proxySpy(
                    (...args) => log.push(...args),
                    prop,
                    v || function() {},
                );
        },
    });

const namedFunction = name => {
    const fn = function() {};
    Object.defineProperty(fn, 'name', { value: name });
    return fn;
};

const KEEP = {};
const MOCK = {};
const SPY = {};

const effectsLogger = scopeFactory => (
    plan,
    { log = newLog(), defaultAction = KEEP, recorder = {} } = {},
) => {
    // to not polute the log with scope creation
    let moduleLoggerEnabled = false;
    let loggerEnabled = false;

    const logger = (...args) => loggerEnabled && log.push(...args);
    const moduleLogger = ([type, ...args]) =>
        moduleLoggerEnabled && type != 'get' && log.push([type, ...args]);

    const scope = scopeFactory(
        proxySpy(moduleLogger, 'module', Object.create(null)),
    );

    // it's a hack for presentation :)
    logger.recorder = recorder;

    for (const id in scope) {
        const action = id in plan ? plan[id] : defaultAction;

        if (action === KEEP) {
            continue;
        }

        if (action === SPY) {
            scope[id] = proxySpy(logger, id, scope[id]);
            continue;
        }

        if (plan[id] === MOCK) {
            delete plan[id];
        }

        if (typeof scope[id] == 'function') {
            if (id in plan && typeof plan[id] != 'function') {
                console.warn(
                    `TestPlan: Mocking a function "${id}" with a non-function mock "${typeof plan[
                        id
                    ]}"`,
                );
            }
            scope[id] = proxySpy(
                logger,
                id,
                plan[id] || namedFunction(`${id}AutoMock`),
            );
            continue;
        }

        if (typeof scope[id] == 'object' && scope[id] !== null) {
            if (id in plan && typeof plan[id] != 'object') {
                console.warn(
                    `TestPlan: Mocking an object "${id}" with a non-object mock "${typeof plan[
                        id
                    ]}"`,
                );
            }
            scope[id] = proxySpy(
                logger,
                id,
                plan[id] || (Array.isArray(scope[id]) ? [] : {}),
            );
            continue;
        }

        // all primitive values stay as they were
    }

    for (const id in plan) {
        if (typeof scope[id] == 'undefined') {
            if (
                typeof plan[id] == 'function' ||
                (typeof plan[id] == 'object' && plan[id] !== null)
            ) {
                scope[id] = proxySpy(logger, id, plan[id]);
            } else {
                // all primitive values stay as they are
                scope[id] = plan[id];
            }
        }
    }

    const effects = () => {
        // to stop loggin after call to effects()
        // to prevent noise from Jest serializer iteration
        loggerEnabled = false;
        moduleLoggerEnabled = false;

        return log;
    };

    loggerEnabled = true;
    moduleLoggerEnabled = true;
    return { scope, effects, m: functionMocker(log) };
};

module.exports = {
    SPY,
    KEEP,
    MOCK,
    effectsLogger,
    isEffectsLoggerLog,
    newLog,
    functionMocker,
};

// auto enable serializers in Jest
require('./jest');
