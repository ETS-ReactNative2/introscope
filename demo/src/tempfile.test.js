import { readFileSync, writeFileSync } from 'fs';
import { introscope } from './tempfile';
import { effectsLogger, SPY } from 'introscope/logger';

const logger = effectsLogger(introscope);

const getRecorder = testName => {
    const resultsFile = `${__dirname}/${testName}.results.json`;

    try {
        const results = JSON.parse(readFileSync(resultsFile));
        return {
            save: () => {},
            playback: true,
            results,
        };
    } catch (_) {}

    return {
        recording: true,
        results: [],
        save() {
            writeFileSync(resultsFile, JSON.stringify(this.results));
        },
    };
};

test('generates a kinda uniq name', async () => {
    const recorder = getRecorder('tempfile');

    const { scope, effects } = logger(
        {
            now: SPY,
            rand: SPY,
        },
        { recorder },
    );

    expect(scope.tempfile()).toMatchSnapshot();

    recorder.save();
});
