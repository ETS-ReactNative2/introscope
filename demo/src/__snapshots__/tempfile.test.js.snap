// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates a kinda uniq name 1`] = `"/var/tmp/file-1537390034475-4313593178224871"`;
